### PR TITLE
Refactor, bug fixes, and improvements

### PR DESCRIPTION
**Pull request by raman325**
## Proposed change
Hi there, I started down this path because:
1) I was having trouble getting the integration to work
2) I wanted to figure out if there was a clean way to remove entities on unload since a lot end up getting left over.

I didn't make it very far on 2 but I did make good progress on 1 and in the middle of it added a bunch of scope creep. I apologize for the size of the PR, but I thought it would be a good way to make this more maintainable going forward. If you see some changes you'd like me to make, or if there are parts of the refactor you would like me to revert, let me know. I haven't changed much of the actual logic as you'll see in the notes below since it all comes from the package templates and the sensors.

NOTE: I have tested this and as far as I can tell, everything is working as expected, but my integration hasn't been working since before my code changes (specifically, the PIN gets set but `PIN Status` shows as `Connecting` forever), so it would be great if someone else could test it

### New Features
- Packages folder gets cleaned up on unload

### Changes
- switch to using relative path for CONF_PATH to reduce changes of error (existing config entries will be updated automatically)

### Bug Fixes
- Switch _generate_package from async to sync because it does I/O on the filesystem
- fix output file not being opened with context manager (so never closed)
- fix sensor creation loop to only create necessary sensors instead of sensors for all code slots
- fix CodeSlotsData sensor (now a DataUpdateCoordinator) logic for OZW integrations

### Code Cleanup/Improvements
- move CodeSlotsData logic into DataUpdateCoordinator since that's what it is being used for
- reduce polling frequency significantly by only polling any time a slot is set or cleared with a backup hourly polling in case updates are made outside of KeyMaster (this will significantly improve battery life over current implementation with the same level of accuracy)
- Update config entry using update method instead of directly manipulating attributes
- use slugify when fixing CONF_NAME
- use os.path.join to create paths for better OS compatibility
- switch to using f-strings
- only attempt cleanup of packages directory if it exists
- always rely on entry.data so that options is only used to capture options updates
- simplify code and reduced repeat logic by moving things into common functions
- raise Exceptions in services when there is a failure (this will be shown in UI)
- use more list comprehensions
- switch while loops to for loops
- added typing
- add venv to .gitignore
- add requirements_dev.txt
- improve debug messaging

### TODO (in future PRs)
- On options update, clean up no longer used entities
- On removal, cleanup all entities that were created from packages

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:

### CODE REVIEW

File: custom_components/keymaster/const.py [hunk 1]
```diff
@@ -1,15 +1,13 @@
 DOMAIN = "keymaster"
 VERSION = "0.0.45"
-ISSUE_URL = "https://github.com/FutureTense/keymaster"
+ISSUE_URL = "https://github.com/FutureTense/keypaster"
```
Comment (firstof9) [inline custom_components/keymaster/const.py]: Typo here

**Review by firstof9 — CHANGES_REQUESTED** [review]

File: custom_components/keymaster/const.py (hunk 1 above)
Comment (raman325) [reply custom_components/keymaster/const.py]: thanks, good catch!

**Comment by firstof9** [comment]
These changes look good to me. I like the changes auto updating the code slots data myself.
Guess I won't have to do the DataCoordinator implimentation myself either 🎉

**Review by firstof9 — APPROVED** [review]

File: custom_components/keymaster/__init__.py [hunk 2]
```diff
@@ -35,327 +46,331 @@
...
                 _LOGGER.error(
                     "Error calling ozw.clear_usercode service call: %s", str(err)
                 )
-                # pass
 
-            servicedata = {
-                ATTR_ENTITY_ID: entity_id,
-                ATTR_CODE_SLOT: code_slot,
-            }
+            servicedata[ATTR_ENTITY_ID] = entity_id
```
Comment (firstof9) [inline custom_components/keymaster/__init__.py]: I think this actually needs to be the code slot. Since previously we altered it to slot 999.

**Review by firstof9 — CHANGES_REQUESTED** [review]

File: custom_components/keymaster/__init__.py (hunk 2 above)
Comment (FutureTense) [reply custom_components/keymaster/__init__.py]: @firstof9 I’m not sure what you mean

Comment (raman325) [reply custom_components/keymaster/__init__.py]: I see the problem, pushing a fix shortly - when I had originally written this, we made one service call for both `ozw` and `zwave` so I could set the code slot before the conditional. Now that we are making two service calls in `ozw` it's better to just set the full payload right before the service call.

**Review by firstof9 — APPROVED** [review]
Changes look good now 👍

**Pull request merged**